// @google/clasp/tests/functions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("./constants");
var fs_extra_1 = __importDefault(require("fs-extra"));
exports.cleanup = function () {
    fs_extra_1.default.removeSync('.clasp.json');
    fs_extra_1.default.removeSync('.claspignore');
    fs_extra_1.default.removeSync('Code.js');
    fs_extra_1.default.removeSync('appsscript.json');
};
exports.setup = function () {
    fs_extra_1.default.writeFileSync('.clasp.json', constants_1.CLASP_SETTINGS.valid);
    fs_extra_1.default.writeFileSync('appsscript.json', constants_1.TEST_APPSSCRIPT_JSON_WITHOUT_RUN_API);
};
exports.setupWithoutGCPProject = function () {
    fs_extra_1.default.writeFileSync('.clasp.json', constants_1.CLASP_SETTINGS.validWithoutProjectId);
    fs_extra_1.default.writeFileSync('appsscript.json', constants_1.TEST_APPSSCRIPT_JSON_WITHOUT_RUN_API);
};
exports.setupWithRunManifest = function () {
    fs_extra_1.default.writeFileSync('.clasp.json', constants_1.CLASP_SETTINGS.valid);
    fs_extra_1.default.writeFileSync('appsscript.json', constants_1.TEST_APPSSCRIPT_JSON_WITH_RUN_API);
};
exports.rndStr = function () { return Math.random().toString(36).substr(2); };
exports.backupSettings = function () {
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcGlobal)) {
        fs_extra_1.default.copyFileSync(constants_1.CLASP_PATHS.rcGlobal, constants_1.CLASP_PATHS.rcGlobal + "~");
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcLocal)) {
        fs_extra_1.default.copyFileSync(constants_1.CLASP_PATHS.rcLocal, constants_1.CLASP_PATHS.rcLocal + "~");
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.settingsLocal)) {
        fs_extra_1.default.copyFileSync(constants_1.CLASP_PATHS.settingsLocal, constants_1.CLASP_PATHS.settingsLocal + "~");
    }
};
exports.restoreSettings = function () {
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcGlobal + "~")) {
        fs_extra_1.default.renameSync(constants_1.CLASP_PATHS.rcGlobal + "~", constants_1.CLASP_PATHS.rcGlobal);
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcLocal + "~")) {
        fs_extra_1.default.renameSync(constants_1.CLASP_PATHS.rcLocal + "~", constants_1.CLASP_PATHS.rcLocal);
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.settingsLocal + "~")) {
        fs_extra_1.default.renameSync(constants_1.CLASP_PATHS.settingsLocal + "~", constants_1.CLASP_PATHS.settingsLocal);
    }
};
